forms/OpenIDPluginSettingsForm.php

```diff
@@ -0,0 +1,237 @@
+{
+    private const HIDDEN_CHARS = '******';
+
+    /**
+     * OpenIDPluginSettingsForm constructor.
+     */
+    public function __construct(private OpenIDPlugin $plugin)
+    {
+        parent::__construct($plugin->getTemplateResource('settings.tpl'));
+
+        $this->addCheck(new FormValidatorPost($this));
+        $this->addCheck(new FormValidatorCSRF($this));
+    }
+
+    /**
+     * @copydoc Form::initData()
+     */
+    function initData()
+    {
+        $request = Application::get()->getRequest();
+        $contextId = ($request->getContext() == null) ? 0 : $request->getContext()->getId();
+        $settingsJson = $this->plugin->getSetting($contextId, 'openIDSettings');
+        $settings = json_decode($settingsJson, true);
+        $provider = $settings['provider'];
+
+        if ($provider && is_array($provider)) {
+            foreach ($provider as &$prov) {
+                if (!empty($prov['clientId'])) {
+                    $prov['clientId'] = self::HIDDEN_CHARS;
+                }
+                if (!empty($prov['clientSecret'])) {
+                    $prov['clientSecret'] = self::HIDDEN_CHARS;
+                }
+            }
+        }
+        if (isset($settings)) {
+            $this->_data = [
+                'initProvider' => OpenIDPlugin::PUBLIC_OPENID_PROVIDERS,
+                'provider' => $provider,
+                'legacyLogin' => $settings['legacyLogin'] ?? true,
+                'legacyRegister' => $settings['legacyRegister'] ?? true,
+                'disableConnect' => $settings['disableConnect'] ?? false,
+                'hashSecret' => $settings['hashSecret'],
+                'generateAPIKey' => $settings['generateAPIKey'] ?? 0,
+                'providerSync' => $settings['providerSync'] ?? false,
+                'disableFields' => $settings['disableFields'],
+            ];
+        } else {
+            $this->_data = [
+                'initProvider' => OpenIDPlugin::PUBLIC_OPENID_PROVIDERS,
+                'legacyLogin' => true,
+                'legacyRegister' => false,
+                'generateAPIKey' => false,
+            ];
+        }
+        parent::initData();
+    }
+
+    /**
+     * @copydoc Form::readInputData()
+     */
+    function readInputData()
+    {
+        $this->readUserVars(
+            [
+                'provider',
+                'legacyLogin',
+                'legacyRegister',
+                'disableConnect',
+                'hashSecret',
+                'generateAPIKey',
+                'providerSync',
+                'disableFields'
+            ]
+        );
+        parent::readInputData();
+    }
+
+    /**
+     * @copydoc Form::fetch()
+     */
+    public function fetch($request, $template = null, $display = false)
+    {
+        $context = $request->getContext();
+        $redirectURL = $context == null ? $request->getIndexUrl().'/openid/doAuthentication' : $request->getIndexUrl().'/'.$context->getPath(
```

Comment:
Should be using `$request->url` for URL construction of URLs that point into OJS.

Reply:
Changed this 

```
$redirectURL = $context == null ? $request->getIndexUrl().'/openid/doAuthentication' : $request->getIndexUrl().'/'.$context->getPath(
            ).'/openid/doAuthentication';
        $redirectMSURL = $context == null ? $request->getIndexUrl().'/openid/doMicrosoftAuthentication' : $request->getIndexUrl().'/'.$context->getPath(
            ).'/openid/doMicrosoftAuthentication';
```

to this 

```
$basePath = $context == null ? '' : $context->getPath();
$redirectURL = $request->url($basePath, 'openid', 'doAuthentication');
$redirectMSURL = $request->url($basePath, 'openid', 'doMicrosoftAuthentication');
```